rfd/0015-agent-loading.md

```diff
@@ -0,0 +1,91 @@
+---
+authors: Russell Jones (rjones@goteleport.com)
+state: draft
+---
+
+# RFD 15 - Agent loading.
+
+## What
+
+[#3169](https://github.com/gravitational/teleport/issues/3169) raised UX concerns about how `tsh` loads keys into an SSH agent.
+
+* `tsh` will by default attempt to load the users certificate upon login into any running agent. This breaks some some SSH agents like `gpg-agent` that do not support certificates.
+* `tsh` has multiple flags for certificate loading behavior that users have found confusing.
+
+Furthermore, users have legitimate reasons for wanting to control this behavior (hence the flags). They may want to force load a certificate (our automatic detection algorithm may be incorrect) or not (the user does not want to load Teleport certificates into their agent).
+
+## Why
+
+By default `tsh login` should attempt to load the fetched certificate into the agent. If `tsh` thinks the agent does not support SSH certificates, like `gpg-agent`, Teleport should not load the certificate and instead print a warning informing the user the certificate was not loaded and how to force load.
+
+To make complex things possible a `--add-keys-to-agent` (short flag `-k`) flag shuld be added to `tsh` that can be used to control if `tsh` attempts to load a certificate into the running agent or not.
+
+This matches the [AddKeysToAgent](https://man.openbsd.org/ssh_config.5#AddKeysToAgent) option that OpenSSH supports:
+
+```
+AddKeysToAgent
+
+Specifies whether keys should be automatically added to a running ssh-agent(1). If this option is set to yes
+and a key is loaded from a file, the key and its passphrase are added to the agent with the default lifetime,
+as if by ssh-add(1). If this option is set to ask, ssh(1) will require confirmation using the SSH_ASKPASS
+program before adding a key (see ssh-add(1) for details). If this option is set to confirm, each use of the
+key must be confirmed, as if the -c option was specified to ssh-add(1). If this option is set to no, no keys
+are added to the agent. Alternately, this option may be specified as a time interval using the format
+described in the TIME FORMATS section of sshd_config(5) to specify the key's lifetime in ssh-agent(1),
+after which it will automatically be removed. The argument must be no (the default), yes, confirm (optionally
+followed by a time interval), ask or a time interval.
+```
+
+### Details
+
+The `--add-keys-to-agent` (short flag `-k`) should support following values `auto`, `yes`, `no`. The default value should be `auto`.
+
+* `auto` indicates that a best effort attempt will be made to detect the agent, and if it supports SSH certificates, load the certificate into the running agent.
+* `yes` indicates that an attemp will be made to load the certificate into the running agent.
+* `no` indicates that no attempt will be made to load the certificate into the running agent.
```

Comment:
@russjones let's add `only` option to avoid storing keys on disk

https://github.com/gravitational/teleport/issues/4863